### DISCUSSION TITLE
[RFC] Allow payload for GET request too. (The same as POST, PUT, and other requests)

### DISCUSSION THREAD

**Discussion by realshovanshah** [cfug/dio#1336] [Ideas]
### New Issue Checklist

- [x] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found #252 , but it was closed.
- [ ] Make necessary changes for supporting request body in GET request.

As of version 4.0.3, there's no support for adding request body to GET requests. 

This was previously discussed in #252 and I this has been solved yet. This should be an open issue, so I'm reopening this.

I'm in need of for a project. I'll try to send a pull request if I'm free. 
Do let us know if anyone is interested in helping close this issue.

**Comment by wendux** [comment]
Oh... it is **not** a normal operation to carry a body in a GET request, and the browsers do not support it too...

**Comment by realshovanshah** [comment]
Yes, but there's nothing stopping a backend from adding request body to a GET request. Though browsers don't support it, I'd like to argue that many other client libraries do support this, and it's for a reason. 

After further research I found this- Even Elasticsearch makes use the above mentioned feature.
https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-body.html

**Comment by wendux** [comment]
OK..Let's discuss it together，I want to hear more people's thoughts！

**Reply by kuhnroyal** [reply]
I would allow it, have seen many APIs do this.

**Reply by realshovanshah** [reply]
@Daimas said it best

>  I think you should at least let us do it even if you consider it a bad practice, because we may have no control over the server API.

**Reply by Imgkl** [reply]
Yes. +1
> I think you should at least let us do it even if you consider it a bad practice, because we may have no control over the server API.

**Reply by jgoyvaerts** [reply]
We should allow it, some servers use it, and making Dio uncompatible with some servers seems unlogical.
The servers define the rules, we just have to follow them as a client library.

**Reply by kuhnroyal** [reply]
We could add this in a 5.0, and maybe consider that this requires a specific flag to be set on the request.

**Reply by realshovanshah** [reply]
Would love to send a PR if this gets approved.

**Reply by jgoyvaerts** [reply]
> We could add this in a 5.0, and maybe consider that this requires a specific flag to be set on the request.

Why a flag and not just an optional body parameter?

**Reply by mykolamor** [reply]
I've left an issue #1422 before finding this thread, I think all methods should be allowed to have a body if a user passes it in the request.

**Reply by ntta** [reply]
I'm struggling with the issue of a GET method with a body as well. After a while of digging on the internet, I found the answer on StackOverflow. 

Refer to: https://stackoverflow.com/questions/978061/http-get-with-request-body#:~:text=The%20RFC2616%20referenced,was%20deleted

Years ago, it seems to be a bad practice to use body parameters for a GET request. However, now it should be an acceptable/valid practice for building a GET request like that. I just have a quick thought about an e-commerce website with many different types of filters (e.g.: brand, price range, category, seller name, etc.). Should it be more controllable if we have this kind of filter in a body?

As a back-end API dev, we can change it to a POST request. However, in my opinion, if I use a POST request, I want to change/update something in the database. And a GET method in the above example is exactly what I want to use.

I bet this "new" practice is relying on the evolving of data on the internet 😄 

So, last but not least, hope that Dio can have this option in the next update. 

Cheers!

**Reply by Prn-Ice** [reply]
If it works on postman, it should work on dio 😄

**Reply by HenryEffiong** [reply]
Many back-end API still do this use a body parameter for a GET request. Just encountered one that led me all the way here. The flexibility of having this is option in Dio is much welcomed

**Comment by AlexV525** [comment]
This has been implemented.
